modules/sdk/src/main/scala/org/tessellation/sdk/domain/cluster/programs/L0PeerDiscovery.scala

```diff
@@ -0,0 +1,44 @@
+package org.tessellation.sdk.domain.cluster.programs
+
+import cats.MonadThrow
+import cats.effect.Sync
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import scala.util.control.NoStackTrace
+
+import org.tessellation.schema.peer.{L0Peer, P2PContext}
+import org.tessellation.sdk.domain.cluster.programs.L0PeerDiscovery.UnknownL0PeerDiscoveryError
+import org.tessellation.sdk.domain.cluster.storage.L0ClusterStorage
+import org.tessellation.sdk.http.p2p.clients.L0ClusterClient
+
+object L0PeerDiscovery {
+
+  def make[F[_]: Sync](
+    l0ClusterClient: L0ClusterClient[F],
+    l0ClusterStorage: L0ClusterStorage[F]
+  ): L0PeerDiscovery[F] =
+    new L0PeerDiscovery[F](l0ClusterClient, l0ClusterStorage) {}
+
+  sealed trait L0PeerDiscoveryError extends NoStackTrace
+  case object UnknownL0PeerDiscoveryError extends L0PeerDiscoveryError {
+    override def getMessage: String = "Error during L0 peer discovery!"
+  }
+}
+
+sealed abstract class L0PeerDiscovery[F[_]: Sync] private (
+  l0ClusterClient: L0ClusterClient[F],
+  l0ClusterStorage: L0ClusterStorage[F]
+) {
+
+  def discoverFrom(peer: P2PContext): F[Unit] = {
```

Comment:
I'd be against using for comprehension. Instead, you can think about data flow and use locally reasoning. (for comp introduces kinda global scope)

```scala
l0ClusterClient.getPeers.run(peer)
  .map(_.map(L0Peer.fromPeer))
  .flatMap(l0ClusterStorage.addPeers)
  .handleErrorWith(e => UnknownError...(e).raiseError[F, Unit])
```

Reply:
I get what you mean. With this kind of linear flow of transformations it's unnecessary to name the arguments to then just pass them to the next step. I guess that could be the rule exactly, if you are going from one step to the other depending only on the last step's outcome then using for comprehension isn't needed. Though sometimes I would still use syntactic sugar like this if it improves readability. Not the case here though.

Reply:
but history shows that we have a problem with reasoning locally about arguments in for comp :)